diff --git a/node_modules/babel-preset-gatsby/index.js b/node_modules/babel-preset-gatsby/index.js
index 7da0918..aeb9ec8 100644
--- a/node_modules/babel-preset-gatsby/index.js
+++ b/node_modules/babel-preset-gatsby/index.js
@@ -9,7 +9,7 @@ const loadCachedConfig = () => {
 
   if (process.env.NODE_ENV !== `test`) {
     try {
-      pluginBabelConfig = require(path.join(process.cwd(), `./.cache/babelState.json`));
+      pluginBabelConfig = require(path.join(process.env.GATSBY_CACHE_DIR || path.join(process.cwd(), `./.cache`), `babelState.json`));
     } catch (err) {
       if (err.message.includes(`Cannot find module`)) {
         // This probably is being used outside of the Gatsby CLI.
